### PR TITLE
[+] PDF : Introduce new Invoice template

### PR DESCRIPTION
**Pull request by julienbourdeau**

### CODE REVIEW

File: classes/order/OrderDetail.php [hunk 1]
```diff
@@ -414,6 +414,17 @@ public static function getList($id_order)
 		return Db::getInstance()->executeS('SELECT * FROM `'._DB_PREFIX_.'order_detail` WHERE `id_order` = '.(int)$id_order);
 	}
 
+	public function getTaxList()
+	{
+		return self::getTaxList($this->id);
+	}
+	public static function getTaxListStatic($id_order_detail)
```
Comment (jeromenadaud) [inline classes/order/OrderDetail.php]: Space please!

File: classes/pdf/HTMLTemplateInvoice.php [hunk 2]
```diff
@@ -287,6 +336,59 @@ public function getTaxTabContent()
...
+	protected function getTaxBreakdown()
+	{
+		$breakdowns = array(
+			'product_tax' => $this->order_invoice->getProductTaxesBreakdown($this->order),
+			'shipping_tax' => $this->order_invoice->getShippingTaxesBreakdown($this->order),
+			'ecotax_tax' => $this->order_invoice->getEcoTaxTaxesBreakdown(),
+			'wrapping_tax' => $this->order_invoice->getWrappingTaxesBreakdown(),
+		);
+
+		foreach ($breakdowns as $type => $bd) {
```
Comment (jeromenadaud) [inline classes/pdf/HTMLTemplateInvoice.php]: norms !

File: classes/pdf/HTMLTemplateInvoice.php [hunk 3]
```diff
@@ -287,6 +336,59 @@ public function getTaxTabContent()
...
+			if (empty($bd))
+			unset($breakdowns[$type]);
+		}
+
+		if (empty($breakdowns))
+			$breakdowns = false;
+
+		if (isset($breakdowns['product_tax'])) {
```
Comment (jeromenadaud) [inline classes/pdf/HTMLTemplateInvoice.php]: norms !

File: pdf/invoice.discount-tab.tpl [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+*  @copyright  2007-2015 PrestaShop SA
+*  @license    http://opensource.org/licenses/afl-3.0.php  Academic Free License (AFL 3.0)
+*  International Registered Trademark & Property of PrestaShop SA
+*}
+<table width="100%" border="1" cellpadding="2">
+    <tr>
+        <td class="payment center small grey bold" width="60%">{l s='Discount' pdf='true'}</td>
+        <td class="payment right" width="40%">
+            {assign var="total_other" value=($order_invoice->total_paid_tax_incl - $order_invoice->total_products - $total_tax)}
+            {displayPrice currency=$order->id_currency price=$total_other}
```
Comment (jeromenadaud) [inline pdf/invoice.discount-tab.tpl]: escape please, intval and floatval ;-)

File: pdf/invoice.payment-tab.tpl [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+*}
+<table id="payment-tab" width="100%">
+    <tr>
+        <td class="payment center small grey bold" width="44%">{l s='Payment Method' pdf='true'}</td>
+        <td class="payment left white" width="56%">
+            <table width="100%" border="0">
+                {foreach from=$order_invoice->getOrderPaymentCollection() item=payment}
+                    <tr>
+                        <td class="right small">{$payment->payment_method}</td>
+                        <td class="right small">{displayPrice currency=$payment->id_currency price=$payment->amount}</td>
```
Comment (jeromenadaud) [inline pdf/invoice.payment-tab.tpl]: escape !

File: pdf/invoice.tax-tab.tpl [hunk 6]
```diff
@@ -23,108 +23,73 @@
 *  International Registered Trademark & Property of PrestaShop SA
```
Comment (jeromenadaud) [inline pdf/invoice.tax-tab.tpl]: invoice.tax-tabl is also called in HTMLTemplateOrderSlip.php !

**Comment by Nobodaddy** [comment]
Ok, then let me play the devil's advocate! You may improve whatever you want, not only in my view this new invoice template seems to be a dead end street. Please don't take it personally but I wouldn't agree that this new template is a real improvement. 

This post in the German forum puts it in a nutshell (https://www.prestashop.com/forums/topic/444444-ps-161-beta-mit-advancedeucompliance-%E2%80%93-leider-immer-noch-eine-baustelle/#entry2061602). Here a translation of the major issues:

> "I think Prestashop’s invoice solution is a completely wrong way. While others shop systems already  installed an editor for the user, Prestashop seals off the forms increasingly from prior changes. I also think that the new invoice form is even more far from conventional standards equivalent as the pretty botched previous. It looks as if someone has put into it his ambition to compose the information on an A4 sheet to some aesthetic considerations.
> What should those shops (like ours) do, that enclose the invoices or delivery note with the order and need standard forms whose address field fits in the provided envelopes?
> Though the new account form used when printing less toner than the old one, the design is even further away from the usual business standard than the old one. Especially in Germany, customers expect business letters and invoices, which roughly correspond to the valid DIN standard 5800. Here some examples:
http://rechnungen-muster.de/downloads/Rechnung-Muster-1.pdf
http://rechnungen-muster.de/downloads/Rechnung-Muster-2.pdf
> But the Prestashop bill adheres to no default. It’s not only an address field in the right place, but also a lot of redundancy (eg information on the invoice date and -No top right and again in the subject line.) and an overall amateurish design.

> If Prestashop will gain ground in Germany, what we all want, then those freestyle exercises should be mothballed quickly and give way to a professional solution."

One of the German users installed a test server with 1.6.1RC3 here: 
http://prestashop-2268-12515-39982.cloudwaysapps.com/administrator/

User: test@test.de
PW: testtest

Please have a look at the alternative invoice templates that are uploaded here. You may switch in menu
> Orders -> Invoices
Try e.g. "invoice-new" for an invoice confirm not alone to German business standards. I guess you will realize the difference.

**Comment by julienbourdeau** [comment]
Hi @Nobodaddy,

We made few changes based on your feedback, you can have a look at my latest PR #3158. About the _overall amateurish design_ I'm not sure we'll change anything.

Also, if you're interested in, we could publish your design as well and PrestaShop could maintain some sort of _"invoice template library"_.

**Pull request closed without merging**